cpp/src/arrow/dataset/filter.h

```diff
@@ -305,6 +316,36 @@ class ARROW_DS_EXPORT NotExpression final
   Result<std::shared_ptr<DataType>> Validate(const Schema& schema) const override;
 };
 
+/// Explicitly cast an expression to a different type
+class ARROW_DS_EXPORT CastExpression final
+    : public ExpressionImpl<UnaryExpression, CastExpression, ExpressionType::CAST> {
+ public:
+  CastExpression(std::shared_ptr<Expression> operand, std::shared_ptr<DataType> to,
+                 compute::CastOptions options)
+      : ExpressionImpl(std::move(operand)),
+        to_(std::move(to)),
+        options_(std::move(options)) {}
+
+  CastExpression(std::shared_ptr<Expression> operand, std::shared_ptr<Expression> like,
```

Comment:
The other expression is assumed to be evaluated against the same schema? Perhaps add a docstring?

Reply:
will do